src/data/scripts/hullmods/vayra_modular_shields.java

```diff
@@ -116,68 +122,200 @@ public void applyEffectsAfterShipCreation(ShipAPI ship, String id) {
+        // First, try getting generator directly from the cached data
+        Map<FleetMemberAPI, ShipAPI> storedGenerators;
         if (engine.getCustomData().get(STORED_GENERATORS_KEY) instanceof Map) {
-            storedGenerators = (Map<ShipAPI, ShipAPI>) engine.getCustomData().get(STORED_GENERATORS_KEY);
+            storedGenerators = (Map<FleetMemberAPI, ShipAPI>) engine.getCustomData().get(STORED_GENERATORS_KEY);
         } else {
             storedGenerators = new HashMap<>();
         }
-
+        utilLog("storedGenerators.size(): "+storedGenerators.size());
         // find the generator
-        ShipAPI generator = storedGenerators.get(ship);
+        FleetMemberAPI shipsFleetMember = getFleetMember(ship);
+        ShipAPI generator = storedGenerators.get(shipsFleetMember);
         if (generator == null) {
-            for (ShipAPI check : CombatUtils.getShipsWithinRange(ship.getLocation(), 1000)) {
+            // Now try getting generator directly since we have no cached data
+            List<ShipAPI> shipsWithin1000range = CombatUtils.getShipsWithinRange(ship.getLocation(), 1000);
+
+            for (ShipAPI check : shipsWithin1000range) {
                 if (check.getHullSpec() != null
                         && check.getHullSpec().getHullId() != null
                         && check.getHullSpec().getHullId().equals(SHIELD_GENERATOR_ID)
-                        && (ship.equals(check.getParentStation())
-                        || (ship.getParentStation() != null
-                        && ship.getParentStation().equals(check.getParentStation())))) {
+                        && ship.getParentStation() != null
+                        && ship.getParentStation().equals(check.getParentStation())
+                ) {
                     generator = check;
-                    storedGenerators.put(ship, generator);
+                    storedGenerators.put(shipsFleetMember, generator);
+                }
+            }
+
+            // If we still didn't find the generator, try the alternative approach - from any other part
+            if (generator == null) {
+                for (ShipAPI check : shipsWithin1000range) {
+                    if (check.getHullSpec() != null
+                            && check.getHullSpec().getHullId() != null
+                            && check.getHullSpec().getHullId().equals(SHIELD_GENERATOR_ID)
+                            &&
+                            (
+                                    ship.equals(check.getParentStation())
+                                            ||
+                                            (
+                                                    ship.getParentStation() != null
+                                                            && ship.getParentStation().equals(check.getParentStation())
+                                            )
+                            )
+                    ) {
```

Comment:
consider actually extracting these things into variables so it's more readable. this way it's just.... hideous.

Reply:
done. much better.

Reply:
resolving